core/src/toga/documents.py

```diff
@@ -1,95 +1,58 @@
 from __future__ import annotations
 
-import asyncio
-import warnings
+import itertools
+import sys
 from abc import ABC, abstractmethod
+from collections.abc import Iterator
 from pathlib import Path
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Mapping, Sequence
+
+import toga
+from toga import dialogs
+from toga.handlers import overridable, overridden
+from toga.window import MainWindow, Window
 
 if TYPE_CHECKING:
     from toga.app import App
-    from toga.window import Window
 
 
 class Document(ABC):
```

Comment:
Possible simplifications of the Document class:

* `create`, `open`, and `read` will only be called once, in that order. So what if we merged `create` and `read` into a single overridable method that's responsible for both creating the UI and loading a file into it? For consistency with the other classes, the best name for this method would be `open`. The code currently in `open` can then move to the constructor, which would once again take a `path` argument which can be None for a new document.

* Similarly, the overridable `write` method would be renamed to `save`, and the code currently in `save` can move to `DocumentWindow.save`.

* Separately, `Document.show` seems unnecessary, and also incomplete since there's no `Document.hide`. Suggest removing this method and saying that whoever creates the window is also responsible for showing it, just like everywhere else in Toga.

Reply:
> Possible simplifications of the Document class:

I'm not sure these simplifications are actually possible (or, to the extent they're possible, they won't represent an improvement).

`open()` does more than just read the file; it also does path normalisation, error handling, window title handling, and the initial dirty state management. `read()` is then the extension point that the user can implement so that they *only* need to deal with the "get the data from the file and populate the window" portion, and can implement that with known the pre-condition that the file *must* exist. The only restructure of this that I can see would require either:
* every user implementation of a unified `open()` to call `super().open()` and potentially handle the fact that open might raise an exception or not require opening a file at all; or
* pulling the current logic of `open()` into the place where `open()` is being invoked... which means we've got the same code, just wrapped up into a more complex `DocumentWindow.open()` implementation.

The same is true of `save()` and `write()`.

Separating `create()` from `open()` is necessary to support untitled documents. Otherwise, we'd need to modify `open()` to handle `path=None` (or similar) as a "don't open a file" case. Sure, this is *possible*, but it's a bit messy IMHO. It also requires the user to convolve "creating the window" logic with "reading the file" logic (which makes testing a little harder), and makes operations like "refresh the content of the current window in response to changes on disk" harder (an edge case, to be sure, but not a *completely* unreasonable one).

A standalone `show()` is needed to support deferring making the window visible until after it's fully loaded. Without it, when opening a new document, you'd see an untitled window that is then populated with content (which might take time and result in visible redraws); or, in case of a failed open, a temporary window that then disappears.

I can't think of a good use case for a `hide()` method; but for the sake of parity, it's not hard to implement.